src/decompressor.rs

```diff
@@ -0,0 +1,116 @@
+use std::ffi::OsStr;
+use std::io::Read;
+use std::path::Path;
+use std::collections::HashMap;
+use std::process::Command;
+use std::process::Stdio;
+use std::process::ChildStdout;
+use globset::{Glob, GlobSet, GlobSetBuilder};
+
+/// A decompression command, contains the command to be spawned as well as any
+/// necessary CLI args
+struct DecompressionCommand {
+    cmd: &'static str,
+    args: Vec<&'static str>
+}
+
+impl DecompressionCommand {
+    /// Create a new decompress command
+    fn new(cmd: &'static str, args: Vec<&'static str>) -> DecompressionCommand {
+        DecompressionCommand {
+            cmd, args
+        }
+    }
+}
+
+lazy_static! {
+    static ref DECOMPRESSION_COMMANDS: HashMap<&'static str, DecompressionCommand> = {
+        let mut m = HashMap::new();
+        m.insert("gz", DecompressionCommand::new("gunzip", vec!["-c"]));
+        m.insert("bz2", DecompressionCommand::new("bunzip2", vec!["-c"]));
+        m.insert("xz", DecompressionCommand::new("unxz", vec!["-c"]));
+        m.insert("lzma", DecompressionCommand::new("unlzma", vec!["-c"]));
+        m
+    };
+    static ref SUPPORTED_COMPRESSION_FORMATS: GlobSet = {
+        let mut builder = GlobSetBuilder::new();
+        builder.add(Glob::new("*.gz").unwrap());
+        builder.add(Glob::new("*.bz2").unwrap());
+        builder.add(Glob::new("*.xz").unwrap());
+        builder.add(Glob::new("*.lzma").unwrap());
+        builder.build().unwrap()
+    };
+    static ref TAR_ARCHIVE_FORMATS: GlobSet = {
+        let mut builder = GlobSetBuilder::new();
+        builder.add(Glob::new("*.tar.gz").unwrap());
+        builder.add(Glob::new("*.tar.xz").unwrap());
+        builder.add(Glob::new("*.tar.bz2").unwrap());
+        builder.add(Glob::new("*.tgz").unwrap());
+        builder.add(Glob::new("*.txz").unwrap());
+        builder.add(Glob::new("*.tbz2").unwrap());
+        builder.build().unwrap()
+    };
+}
+
+/// Returns a handle to the stdout of the spawned decompression process for
+/// `path`, which can be directly searched in the worker.
+///
+/// If there is any error in spawning the decompression command, then return
+/// `None`, after outputting any necessary debug or error messages.
+pub fn get_reader(path: &Path, no_messages: bool) -> Option<ChildStdout> {
+    if is_tar_archive(path) {
+        debug!("Tar archives are currently unsupported: {}", path.display());
+        None
```

Comment:
In general, I like to keep control flow straight-forward and avoid gratuitous nesting. Early returns are very useful for this. For example, this code could be written as:

```rust
if is_tar_archive(path) {
    debug!(...);
    return None;
}

let extension = ...;
```

That way, you don't need to indent the entire rest of the function.

Reply:
Ah, that makes a lot of sense. I did do this in the `worker.rs` file but missed it here. 👍